### ISSUE TITLE
timeout connecting to redis hosted on aws ElastiCache

### ISSUE THREAD

**Issue by Jun711** [redis/redis-py#1001]
Hi,

I have a redis server hosted by aws using aws ElastiCache. When I try to connect to it using a lambda and on my local machine, it gives 'timeout connecting to server' error. However, my colleague could connect to this redis using another client in PHP. 
I wonder what I did wrong. 

```
try:
        redisClient = redis.Redis(host='***.use1.cache.amazonaws.com', port=6379, db=0, socket_timeout=10)
        print(redisClient)
        try:
            print(redisClient.ping())
            print(redisClient.set('foo','bar'))
            # r.get('foo')
        except Exception as e:
            print('get set error: ', e)
    except Exception as e:
        print('redis err: ', e)

// redisClient Redis<ConnectionPool<Connection<host=***.use1.cache.amazonaws.com,port=6379,db=0>>>

// get set error:  Timeout connecting to server for ping or set
```

Without setting the socket_timeout parameter, it will time out based on redis-py operation timeout:
Error 60 connecting to ***.use1.cache.amazonaws.com:6379. Operation timed out.

**Comment by dejlek** [comment]
I have similar problem - our Redis has 20+ thousand clients (long story why is that so) - however the error is 110. I am going to use redis with `retry_on_timeout=True` and see whether it helps... Here is the exception:

    redis.exceptions.ConnectionError: Error 110 connecting to server.example.com:6379. Connection timed out.

**Comment by 211217613** [comment]
> When I try to connect to it using a lambda and on my **local machine**

- You can only connect to elasticache via a machine running in the same region.

**Comment by grumps** [comment]
Yes you need to be in the same region. Also test with the redis-cli too, to rule out firewall rules, routing etc.

**Comment by dejlek** [comment]
This is happening randomly. Is there are way to tell redis-py to reconnect on connection timeouts with some backoff strategy?

**Comment by theodesp** [comment]
That's a good PR. Retry with backoff.

**Comment by andymccurdy** [comment]
@dejlek There's a `retry_on_timeout` option in redis-py that will attempt to reconnect and reissue a command when a timeout occurs. If the command again times out, the timeout is then raised to the app to handle.

However, be careful with this setting. There is a chance for data duplication or data loss in very specific scenarios. For example, if you send an `RPUSH` command successfully but time out on reading the reply from `RPUSH`, retrying the command could cause the data to be duplicated in the list.

**Comment by john-feusi-neuro** [comment]
I ran into this issue as well, but from a Lambda. For me, there were a few problems that had to be ironed out
- The lambda needs [VPC permissions](https://docs.aws.amazon.com/lambda/latest/dg/vpc-ec.html#vpc-ec-create-iam-role).
- The ElastiCache security group needs an inbound rule from the Lambda security group that allows communication on the Redis port. I thought they could just be in the same security group.
- And the real kicker: I had turned on encryption in-transit. This meant that I needed to pass `redis.RedisClietn(... ssl=True)`. The redis-py page mentions that `ssl_cert_reqs` needs to be set to `None` for use with ElastiCache, but that didn't seem to be true in my case. I did however need to pass `ssl=True`.

It makes sense that `ssl=True` needed to be set but the connection was just timing out so I went round and round trying to figure out what the problem with the permissions/VPC/SG setup was.

**Comment by soumilshah1995** [comment]
I am running into similar issue

**Comment by konstankinollc** [comment]
Please make sure your Redis placement is in the same availability zone as your server. Multi-AZ will allow you to access your server from multiple AZs.

**Comment by omaraltayyan** [comment]
> Please make sure your Redis placement is in the same availability zone as your server. Multi-AZ will allow you to access your server from multiple AZs.

I am sorry this is not true, Multi-AZ is for increasing the availability of your setup, so if something happens to one of your redis instances the other one will take its place.

You can connect to an ElasticCache instance if it's in the same region, no need to worry about AZs for that.

**Comment by samitpal** [comment]
Retry with backoff like [redis-rb](,https://github.com/redis/redis-rb#reconnections )might be useful I guess.

**Comment by ronyeh-lb** [comment]
hey, we are facing same timeouts issues on EXISTS cmd, does anyone have any proposal ?

**Comment by dejlek** [comment]
@chayim , do you by any chance know how to make Celery use `retry_on_timeout` setting by any chance?

**Comment by LuanNVVnB** [comment]
you need to reconfigure the client and on redislab you need info in connect #

// create a client instance
const client = redis.createClient({
    password: PASSREDIS,
    socket: {
        host: URLREDIS,
        port: PORTREDIS,
    },
});

**Comment by kokani** [comment]
> I ran into this issue as well, but from a Lambda. For me, there were a few problems that had to be ironed out
> 
> * The lambda needs [VPC permissions](https://docs.aws.amazon.com/lambda/latest/dg/vpc-ec.html#vpc-ec-create-iam-role).
> * The ElastiCache security group needs an inbound rule from the Lambda security group that allows communication on the Redis port. I thought they could just be in the same security group.
> * And the real kicker: I had turned on encryption in-transit. This meant that I needed to pass `redis.RedisClietn(... ssl=True)`. The [redis-py](https://pypi.org/project/redis/) page mentions that `ssl_cert_reqs` needs to be set to `None` for use with ElastiCache, but that didn't seem to be true in my case. I did however need to pass `ssl=True`.
> 
> It makes sense that `ssl=True` needed to be set but the connection was just timing out so I went round and round trying to figure out what the problem with the permissions/VPC/SG setup was.

Thanks John! I had encryption turned on as well and the ssl=True solved it for me. Much appreciated!

**Comment by tomaszczechowski** [comment]
In my case, connection via TLS solved the problem since the encryption was enabled. This means that I should use instead of `redis://` => `rediss://`.

**Comment by petyaslavova** [comment]
Hi, this issue hasn’t been updated in several months. Since there are multiple proposed configuration changes, I’ll close it for now. Please feel free to reopen it if further assistance is needed.
